handle multiple returns properly in boxing wrappers

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42438 add Quantizer support to IValue
* **#42437 handle multiple returns properly in boxing wrappers**

1. Previously, a wrapper for a boxed kernel of an op with multiple return values would expect the kernel to have pushed a single IValue containing a tuple of return values. This is incorrect; boxed kernels should push each return value individually. This PR provides the correct plumbing. [important part](https://github.com/pytorch/pytorch/pull/42437/files#diff-63afcafe5c0813e43103b1a242566847R163-R236) (note: for some reason, 
GH doesn't scroll you down to the selection - search for "PopResult")

2. We also take the opportunity to change some asserts into errors, if the conditions may be triggered by user code (e.g. a boxed kernel pushing the incorrect number of return values onto the stack).

3. Added tests for functional (i.e., not in-place/out-place) ops with multiple return values.

Differential Revision: [D22894191](https://our.internmc.facebook.com/intern/diff/D22894191)

Commit:
handle multiple returns properly in boxing wrappers

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -262,9 +304,11 @@ struct BoxedKernelWrapper<
     torch::jit::push(stack, std::forward<Args>(args)...);
 
     (*boxed_kernel_func)(functor, opHandle, &stack);
-    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
```
Why TORCH_CHECK? Do you think this could happen when it is not a PyTorch bug but a user error?

bhosmer replied:
yeah, I mean all these boxed kernels are on the honor system, right? any time a user-supplied kernel pushes the wrong number of return values, it's going to end up here. (E.g. this happened to Richard [here](https://gist.github.com/zou3519/90428b51bf4b51285ac9c06de5c8f98e) - in this case it turned out he was right and the boxing code was wrong, but the effect would have been the same in the reverse case)

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -148,17 +149,55 @@ using has_quantizer_arg =
+  }
+private:
+  // note: this has been moved into its own helper only to avoid a parse error on `indices` otherwise.
+  // I'm sure there's an incantation that slips it past the parser but eh
```
This way of expanding indices is a common pattern. Don't think there's an incantation to avoid the helper.

bhosmer replied:
Ugh templates are such a hack

bhosmer replied:
😬

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -223,7 +263,7 @@ struct BoxedKernelWrapper<
 
     (*boxed_kernel_func)(functor, opHandle, &stack);
-    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
+    TORCH_CHECK(
```
Did you measure if the removal of DEBUG_ONLY has a perf impact?

bhosmer replied:
I didn't, and I expect it does have some marginal impact, but (a) I looked around and kernels seem to use `TORCH_CHECK` pretty liberally for similar purposes (e.g. argument checking), so it seems idiomatic... and (b) the alternative seems like bad user experience (stack underflows or silent throwing away values).

Also maybe worth noting that this codepath won't get exercised by boxed fallbacks or profiling, the two use cases I can think of where the speed probably matters most.

smessmer commented on aten/src/ATen/core/boxing/impl/boxing.h:
```diff
@@ -223,7 +263,7 @@ struct BoxedKernelWrapper<
     torch::jit::push(stack, std::forward<OtherArgs>(otherArgs)...);
```
Does this handle ArrayRef arguments correctly or do they somehow not happen in calls here?

bhosmer replied:
There's definitely no logic to avoid ArrayRef args here (or in the other specializations - they all do this - are you thinking this one has a particular issue?) will follow up offline...

Commit:
Update on "handle multiple returns properly in boxing wrappers"


1. Previously, a wrapper for a boxed kernel of an op with multiple return values would expect the kernel to have pushed a single IValue containing a tuple of return values. This is incorrect; boxed kernels should push each return value individually. This PR provides the correct plumbing. [important part](https://github.com/pytorch/pytorch/pull/42437/files#diff-63afcafe5c0813e43103b1a242566847R163-R236) (note: for some reason, 
GH doesn't scroll you down to the selection - search for "PopResult")

2. We also take the opportunity to change some asserts into errors, if the conditions may be triggered by user code (e.g. a boxed kernel pushing the incorrect number of return values onto the stack).

3. Added tests for functional (i.e., not in-place/out-place) ops with multiple return values.

Differential Revision: [D22894191](https://our.internmc.facebook.com/intern/diff/D22894191)

[ghstack-poisoned]

Commit:
Update on "handle multiple returns properly in boxing wrappers"


1. Previously, a wrapper for a boxed kernel of an op with multiple return values would expect the kernel to have pushed a single IValue containing a tuple of return values. This is incorrect; boxed kernels should push each return value individually. This PR provides the correct plumbing. [important part](https://github.com/pytorch/pytorch/pull/42437/files#diff-63afcafe5c0813e43103b1a242566847R163-R236) (note: for some reason, 
GH doesn't scroll you down to the selection - search for "PopResult")

2. We also take the opportunity to change some asserts into errors, if the conditions may be triggered by user code (e.g. a boxed kernel pushing the incorrect number of return values onto the stack).

3. Added tests for functional (i.e., not in-place/out-place) ops with multiple return values.

Differential Revision: [D22894191](https://our.internmc.facebook.com/intern/diff/D22894191)

[ghstack-poisoned]
